File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
      ListNode temp=head;
      int size=0;
        while(temp!=null)
        {
            size++;
            temp=temp.next;
        }
        return reverse(head,k,size);

    }
    static ListNode reverse(ListNode head,int k,int size)
    {
        ListNode curr=head;
        // ListNode next=head;
        ListNode prv=null;
         ListNode next=head;
        int count=0;
        while(curr!=null && count<k)
        {
            next=next.next;
            curr.next=prv;
            prv=curr;
            curr=next;
            count++;
            size--;
        }
        if(next!=null && size>=k)
        {
            ListNode rest_head=reverse(curr,k,size);
            head.next=rest_head;
        }
        if(size<k)
        {
            head.next=curr;
            return prv;
        }
        return prv;
    }

}

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseKGroup(self, head, k):
        start=head
        end=head
        next_start=None
        count=0
        while(end!=None):
            count+=1
            if k==count:
                if end.next!=None:
                    next_start=end.next
                else:
                    next_start=None    
                end.next=None
                # prv=start
                head2=self.reverse(start)
                if start==head:
                    head=head2
                else:
                    prv.next=head2
                if next_start!=None:    
                    prv=start    
                    start=next_start
                    end=next_start
                    count=1

            if next_start!=None:
                prv.next=next_start
            end=end.next

        return head
    
    def reverse(self,head):
        curr=head
        next=head
        prv=None
        while(next!=None):
            next=next.next
            curr.next=prv
            prv=curr
            curr=next        
        return prv    